### ISSUE TITLE
aws_kms_grant EOF issue when permission is missing

### ISSUE THREAD

**Issue by davidgoate** [terraform-providers/terraform-provider-aws#5062]
# Environment

```bash
terraform -version
Terraform v0.11.7
+ provider.aws v1.24.0
```

OSX 10.13.5

# Issue
Given the following plan that has already been applied:

```yaml
resource "aws_kms_key" "sqs_key" {
  description = "Key for encrypting SQS messages"
  deletion_window_in_days = 10
}

resource "aws_kms_alias" "sqs_key" {
  name = "alias/terraform-sqs"
  target_key_id = "${aws_kms_key.sqs_key.key_id}"
}

resource "aws_iam_role" "sns_sqs_kms" {
  name = "terraform-sns-sqs-kms"
  assume_role_policy = "${data.aws_iam_policy_document.sns_sqs_kms_iam_policy.json}"
}

data "aws_iam_policy_document" "sns_sqs_kms_iam_policy" {
  statement {
    actions = [
      "sts:AssumeRole"
    ]
    principals = {
      type = "Service"
      identifiers = [
        "sns.amazonaws.com"
      ]
    }
  }
}
```

then adding:

```yaml
resource "aws_kms_grant" "sns_sqs_key_grant" {
  name = "sns_sqs_key_grant"
  key_id = "${aws_kms_key.sqs_key.key_id}"
  grantee_principal = "${aws_iam_role.sns_sqs_kms.arn}"
  operations = [
    "GenerateDataKey",
    "GenerateDataKeyWithoutPlaintext",
    "Decrypt"
  ]
}
```

The output of `terraform apply` is:

```
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  + aws_kms_grant.sns_sqs_key_grant
      id:                    <computed>
      grant_id:              <computed>
      grant_token:           <computed>
      grantee_principal:     "arn:aws:iam::<my-account-id>:role/terraform-sns-sqs-kms"
      key_id:                "<my-key-id>"
      name:                  "sns_sqs_key_grant"
      operations.#:          "3"
      operations.1237510779: "Decrypt"
      operations.2524777559: "GenerateDataKeyWithoutPlaintext"
      operations.3592719599: "GenerateDataKey"
      retire_on_delete:      "false"


Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

aws_kms_grant.sns_sqs_key_grant: Creating...
  grant_id:              "" => "<computed>"
  grant_token:           "" => "<computed>"
  grantee_principal:     "" => "arn:aws:iam::<my-account-id>:role/terraform-sns-sqs-kms"
  key_id:                "" => "<my-key-id>"
  name:                  "" => "sns_sqs_key_grant"
  operations.#:          "" => "3"
  operations.1237510779: "" => "Decrypt"
  operations.2524777559: "" => "GenerateDataKeyWithoutPlaintext"
  operations.3592719599: "" => "GenerateDataKey"
  retire_on_delete:      "" => "false"

Error: Error applying plan:

1 error(s) occurred:

* aws_kms_grant.sns_sqs_key_grant: 1 error(s) occurred:
* aws_kms_grant.sns_sqs_key_grant: unexpected EOF
```

# Root Cause
With debug enabled I see:

```bash
2018/07/03 18:06:38 [DEBUG] plugin: waiting for all plugin processes to complete...
2018-07-03T18:06:38.947+0100 [WARN ] plugin: error closing client during Kill: err="connection is shut down"
Error: Error applying plan:

1 error(s) occurred:

2018-07-03T18:06:38.354+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 2018/07/03 18:06:38 [DEBUG] [aws-sdk-go] {"__type":"AccessDeniedException","Message":"User: arn:aws:iam::xyz:user/xyz is not authorized to perform: kms:ListGrants on resource: arn:aws:kms:eu-west-1:xyz:key/xyz"}
...
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: panic: runtime error: invalid memory address or nil pointer dereference
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x2b1ac2e]
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: goroutine 1349 [running]:
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.findKmsGrantById(0xc420534c50, 0xc4202b2690, 0x24, 0xc4202b26b5, 0x40, 0x0, 0xc42065cf38, 0x2, 0x2)
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_kms_grant.go:401 +0x22e
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.findKmsGrantByIdWithRetry.func1(0x1039d8f)
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_kms_grant.go:328 +0x72
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.Retry.func1(0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/wait.go:22 +0x75
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.(*StateChangeConf).WaitForState.func1(0xc420838c60, 0xc420546fc0, 0xc420c10380, 0xc42001f320, 0xc420647690, 0xc420647678)
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/state.go:103 +0x4e0
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: created by github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource.(*StateChangeConf).WaitForState
2018-07-03T18:06:38.357+0100 [DEBUG] plugin.terraform-provider-aws_v1.24.0_x4: 	/opt/teamcity-agent/work/222ea50a1b4f75f4/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/resource/state.go:80 +0x1d0
2018/07/03 18:06:38 [TRACE] root: eval: *terraform.EvalWriteState
2018/07/03 18:06:38 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2018/07/03 18:06:38 [TRACE] root: eval: *terraform.EvalIf
2018/07/03 18:06:38 [TRACE] root: eval: *terraform.EvalWriteState
2018/07/03 18:06:38 [TRACE] root: eval: *terraform.EvalWriteDiff
2018/07/03 18:06:38 [TRACE] root: eval: *terraform.EvalApplyPost
2018/07/03 18:06:38 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aws_kms_grant.sns_sqs_key_grant: unexpected EOF
2018/07/03 18:06:38 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aws_kms_grant.sns_sqs_key_grant: unexpected EOF
2018/07/03 18:06:38 [TRACE] [walkApply] Exiting eval tree: aws_kms_grant.sns_sqs_key_grant
2018/07/03 18:06:38 [TRACE] dag/walk: upstream errored, not walking "provider.aws (close)"
2018/07/03 18:06:38 [TRACE] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2018/07/03 18:06:38 [TRACE] dag/walk: upstream errored, not walking "root"
```

I think it'd be nicer if terraform could give a nice error without me turning on DEBUG 👍

**Comment by bflad** [comment]
Thanks for reporting this @davidgoate and sorry you ran into this unexpected behavior! I have submitted a bug fix pull request that should return an error in this situation instead of crashing: #5063

**Comment by bflad** [comment]
The fix for the resource to return errors in this scenario should be handled in master now and will release with version 1.27.0 of the AWS provider, likely middle of next week.

**Comment by davidgoate** [comment]
👍 thanks!

**Comment by bflad** [comment]
This has been released in [version 1.27.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1270-july-11-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
